practice/flutter/flutter/logic.py

```
@@ -0,0 +1,239 @@
+    except models.User.DoesNotExist:
+        raise KeyError(err_str)
+    except IndexError:
+        raise KeyError(err_str)
+
+    return user
+
+
+def get_fluttuser_from_stduser(std_user: User):
+    """ Ensure that FluttUser is created """
+    try:
+        fu = std_user.flutteruser
+    except models.FlutterUser.DoesNotExist:
+        fu = models.FlutterUser(user=std_user)
+        fu.save()
+
+    return fu
+
+
+def preprocess_flutt(text: str):
+    """ Pre-process text minimally before saving
+
+    :param text
+    :return: str - filtered text
+    >>> preprocess_flutt(' This  is <em>not</em> the Markdown we are looking for')
+    'This is not the Markdown we are looking for'
+    """
+    processed = re.sub(r'\s+', ' ', text.strip())  # type: str
+    processed = re.sub(r'<[^>]+>', '', processed)
+    return processed
+
+
+def validate_flutt(text: any) -> any:
+    """ Validate text for saving
+
+    :raise: ValueError if invalid
+    :returns: processed text if valid, False if invalid, None if None passed
+    >>> validate_flutt(' This flutt has issues, but it  is valid.  ')
+    'This flutt has issues, but it is valid.'
+    >>> validate_flutt('<script src="http://evil.website/">')
+    Traceback (most recent call last):
+    ...
+    ValueError: invalid_flutt
+    """
+    if text is None:
+        flutt = text
+    else:
+        flutt = preprocess_flutt(text)
+
+        if len(flutt) == 0:
+            raise ValueError('invalid_flutt')
+
+    return flutt
+
+
+def save_flutt(text: str, user: User) -> models.Flutt:
+    """ Save Flutt data
+
+    :raise: ValueError if invalid text
+    :returns: models.Flutt if saved, None if None passed
+    >>> user = User.objects.create_user('foo', password='letmein')
+    >>> flutt = save_flutt("barbaz", user)
+    >>> flutt.user.username() == 'foo' and flutt.text == 'barbaz'
+    True
+    """
+    flutt_text = validate_flutt(text)
+
+    if flutt_text is None:
+        flutt = None
```

Similar feedback here about silently passing `None`.